### PR TITLE
feat: implement context handler to store HTTP request and tracing information

### LINKED ISSUE #750: structured logging: Implement  request context holder
Add a capability of managing multiple request contexts per-thread. The capability should capture relevant information as `com.google.cloud.logging.HttpRequest` instance or using a builder that populates the minimal set of fields including:
- URI (full URL including query parameters)
- Http method
- Referrer
- Sender IP
- Receiver IP

Parent: #687

### PR DESCRIPTION
**Pull request by minherz**
Implement support for setting current context to capture Http request and tracing information of the current context.
It provides per-thread context handling and supports enter/exit scope to avoid re-setting the context multiple time in event of re-entering the scope.
The context supports setting trace and span ids directly and parsing the ids from `X-CLOUD-TRACE` and `traceparent` Http header values.

Fixes #750

### CODE REVIEW

**Comment by minherz** [comment]
Note that PR also includes pulled env-tests-logging submodule.

**Review by losalex — CHANGES_REQUESTED** [review]
Good job! Lets just touch base about some issues I mentioned in PR

File: google-cloud-logging/src/main/java/com/google/cloud/logging/context/ContextHandler.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+/** Class provides a per-thread storage of the {@see Context} instances. */
+public class ContextHandler {
+  private static final ThreadLocal<Context> _context = new ThreadLocal<>();
+
+  public static Context getCurrentContext() {
+    return _context.get();
+  }
+
+  public static void setCurrentContext(Context context) {
+    _context.set(context);
```
[the remark this answers is not in the document]
Comment (minherz) [reply google-cloud-logging/src/main/java/com/google/cloud/logging/context/ContextHandler.java:28]: `InheritableThreadLocal`replicates the value to the child thread when the child thread is created. It does not work well when application uses thread pools. In that case the data is copied only if the thread pool decides to create the thread. However, the data remains there after the thread is used first time. The following users of the thread will get confused with the outdated data.

File: google-cloud-logging/src/main/java/com/google/cloud/logging/context/RequestContextScope.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.logging.context;
+
+public class RequestContextScope {
+  private static final ThreadLocal<Integer> contextDepth = new ThreadLocal<>();
```
[the remark this answers is not in the document]
Comment (minherz) [reply google-cloud-logging/src/main/java/com/google/cloud/logging/context/RequestContextScope.java:20]: I am not sure what can be a source of confusion that requires explanation. The code uses the variable as a counter to perform a certain action when it is 0 or null.
Please, explain the reason that requires additional documentation or the comment that should eliminate the confusion.

File: google-cloud-logging/src/main/java/com/google/cloud/logging/context/Context.java [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+          if (!versionFormat.equals("00")) {
+            throw new IllegalArgumentException("Not supporting versionFormat other than \"00\"");
+          }
+        } else {
+          throw new IllegalArgumentException(
+              "Invalid format of the header value. Expected \"00-traceid-spanid-arguments\"");
+        }
+        if (fields.length > 1) {
+          String traceId = fields[1];
+          if (traceId.length() > 0) {
```
[the remark this answers is not in the document]
Comment (minherz) [reply google-cloud-logging/src/main/java/com/google/cloud/logging/context/Context.java:160]: You propose to replace the test for zero length with the test for zero length _AND_ `null`.
I respectfully disagree with the statement that condition of "string length greater than zero" is not readable enough.
Also, there is no reason to test the string for `null`. The current code is based on the known behavior of the `String.split()` method like in your suggestion with using `cloudTrace.split(";")[0]`.

File: google-cloud-logging/src/main/java/com/google/cloud/logging/context/Context.java [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+     * Sets the trace id and span id values by parsing the string which represents xCloud Trace
+     * Context. The Cloud Trace Context is passed as {@code x-cloud-trace-context} header (can be in
+     * Pascal case format). The string format is <code>TRACE_ID/SPAN_ID;o=TRACE_TRUE</code>.
+     *
+     * @see <a href="https://cloud.google.com/trace/docs/setup#force-trace">Cloud Trace header
+     *     format.</a>
+     */
+    public Builder loadCloudTraceContext(String cloudTrace) {
+      if (cloudTrace != null) {
+        int split = cloudTrace.indexOf(';');
```
[the remark this answers is not in the document]
Comment (minherz) [reply google-cloud-logging/src/main/java/com/google/cloud/logging/context/Context.java:112]: will do.

File: google-cloud-logging/src/main/java/com/google/cloud/logging/context/Context.java [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+    /**
+     * Sets the IP address (IPv4 or IPv6) of the origin server that the request was sent to.
+     * Examples: {@code 192.168.1.1}, {@code FE80::0202:B3FF:FE1E:8329}.
+     */
+    public Builder setServerIp(String serverIp) {
+      this.requestBuilder.setServerIp(serverIp);
+      return this;
+    }
+
+    /** Sets the string as a trace id value. */
```
[the remark this answers is not in the document]
Comment (minherz) [reply google-cloud-logging/src/main/java/com/google/cloud/logging/context/Context.java:90]: I do not think it should be here. Notions of trace and span are generic. There is no reason to document them in opposite to Cloud Trace or W3C Trace Parent contexts.

File: google-cloud-logging/src/main/java/com/google/cloud/logging/context/Context.java [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+ */
+
+package com.google.cloud.logging.context;
+
+import com.google.cloud.logging.HttpRequest;
+import com.google.cloud.logging.HttpRequest.RequestMethod;
+import com.google.common.base.MoreObjects;
+import java.util.Objects;
+
+/** Class to hold context attributes including information about {@see HttoRequest} and tracing. */
```
[the remark this answers is not in the document]
Comment (minherz) [reply google-cloud-logging/src/main/java/com/google/cloud/logging/context/Context.java:24]: will do

File: google-cloud-logging/src/main/java/com/google/cloud/logging/context/Context.java [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+        if (fields.length > 0) {
+          String versionFormat = fields[0];
+          if (!versionFormat.equals("00")) {
+            throw new IllegalArgumentException("Not supporting versionFormat other than \"00\"");
+          }
+        } else {
+          throw new IllegalArgumentException(
+              "Invalid format of the header value. Expected \"00-traceid-spanid-arguments\"");
+        }
+        if (fields.length > 1) {
```
[the remark this answers is not in the document]
Comment (minherz) [reply google-cloud-logging/src/main/java/com/google/cloud/logging/context/Context.java:158]: you mean the nesting at line 162?
will do.

File: google-cloud-logging/src/main/java/com/google/cloud/logging/context/Context.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (losalex) [reply google-cloud-logging/src/main/java/com/google/cloud/logging/context/Context.java:158]: It is a good practice to emphasize data type checks properly - helps readability (thats why `Strings.isNullOrEmpty(traceId)` was developed :-)). Also in current context call you operate on array and string and both use `length()` call which increases cognitive load on a reader. Correct me if I wrong, you can use isEmpty() is null check is not required.

File: google-cloud-logging/src/main/java/com/google/cloud/logging/context/RequestContextScope.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (losalex) [reply google-cloud-logging/src/main/java/com/google/cloud/logging/context/RequestContextScope.java:20]: I believe that good sample in a comment showing how it should be used is a good idea - this is public API and users should see suggestions when and how this functionality should be integrated. To be more specific, except of the tests, I do not see anywhere in this code a comment showing the usage of this feature which I believe should exist. Perhaps this is not the right place for a sample, but I think we should have it somewhere.

File: google-cloud-logging/src/main/java/com/google/cloud/logging/context/ContextHandler.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (losalex) [reply google-cloud-logging/src/main/java/com/google/cloud/logging/context/ContextHandler.java:28]: How child thread will get those attributes then?

File: google-cloud-logging/src/main/java/com/google/cloud/logging/context/Context.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (minherz) [reply google-cloud-logging/src/main/java/com/google/cloud/logging/context/Context.java:158]: will do

File: google-cloud-logging/src/main/java/com/google/cloud/logging/context/RequestContextScope.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (minherz) [reply google-cloud-logging/src/main/java/com/google/cloud/logging/context/RequestContextScope.java]: decided to remove the class.

File: google-cloud-logging/src/main/java/com/google/cloud/logging/context/ContextHandler.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (minherz) [reply google-cloud-logging/src/main/java/com/google/cloud/logging/context/ContextHandler.java]: configurable initialization using configuration property is implemented to allow to developer defining the way context will be held.

File: google-cloud-logging/src/main/java/com/google/cloud/logging/LoggingConfig.java [hunk 8]
```diff
@@ -134,7 +147,7 @@ private Formatter getFormatterProperty(String name, Formatter defaultValue) {
     String stringFilter = getProperty(name);
     try {
       if (stringFilter != null) {
-        Class clz = ClassLoader.getSystemClassLoader().loadClass(stringFilter);
+        Class<?> clz = ClassLoader.getSystemClassLoader().loadClass(stringFilter);
```
Comment (losalex) [inline google-cloud-logging/src/main/java/com/google/cloud/logging/LoggingConfig.java:150]: nit: Is there any change to `loadClass` function in this PR? I just wonder why this change is here...

File: google-cloud-logging/src/main/java/com/google/cloud/logging/TraceLoggingEnhancer.java [hunk 9]
```diff
@@ -19,16 +19,9 @@
...
-  public TraceLoggingEnhancer() {
-    traceIdLabel = TRACE_ID;
-  }
-
-  public TraceLoggingEnhancer(String prefix) {
-    traceIdLabel = (prefix != null) ? prefix + TRACE_ID : TRACE_ID;
-  }
+  public TraceLoggingEnhancer(String prefix) {}
 
   private static final ThreadLocal<String> traceId = new ThreadLocal<>();
```
Comment (losalex) [inline google-cloud-logging/src/main/java/com/google/cloud/logging/TraceLoggingEnhancer.java:26]: Are we supposed to use same approach as one used in `ContextHandler` (e.g. use ThreadLocal or InheritableThreadLocal based on config)?

File: google-cloud-logging/src/main/java/com/google/cloud/logging/ContextHandler.java [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+  private static ThreadLocal<Context> initContextHolder() {
+    LoggingConfig config = new LoggingConfig(ContextHandler.class.getName());
+    if (config.getUseInheritedContext()) {
+      return new InheritableThreadLocal<>();
+    } else {
+      return new ThreadLocal<>();
+    }
+  }
+
+  public Context getCurrentContext() {
```
Comment (losalex) [inline google-cloud-logging/src/main/java/com/google/cloud/logging/ContextHandler.java:39]: I wonder if we should make it generic by turning this into template to be reused with other objects (not only Context). Seems that `TraceLoggingEnhancer` below might benefit from it as well

**Review by losalex — COMMENTED** [review]
Overall looks good.
The only thing I wondered if we can turn ContextHandler to be a template class to hold any context and config class name could be transferred as parameter to `initContextHolder()` - let me know WDYT

**Review by losalex — APPROVED** [review]

File: google-cloud-logging/src/main/java/com/google/cloud/logging/ContextHandler.java (hunk 10 above)
Comment (minherz) [reply google-cloud-logging/src/main/java/com/google/cloud/logging/ContextHandler.java:39]: I think we should focus on supporting "our" context since this is what this feature is designed to be. Providing template means we include a generic-purpose implementation into logging library code.
If there will be a need for additional information for the context, the `Context` class can be extended to include additional data fields.

File: google-cloud-logging/src/main/java/com/google/cloud/logging/LoggingConfig.java (hunk 8 above)
Comment (minherz) [reply google-cloud-logging/src/main/java/com/google/cloud/logging/LoggingConfig.java:150]: This change eliminates a warning. Nothing else.

File: google-cloud-logging/src/main/java/com/google/cloud/logging/TraceLoggingEnhancer.java (hunk 9 above)
Comment (minherz) [reply google-cloud-logging/src/main/java/com/google/cloud/logging/TraceLoggingEnhancer.java:26]: I can undo the change since this file is not part of the implementation. The changes here are meant to eliminate compilation warnings only.
